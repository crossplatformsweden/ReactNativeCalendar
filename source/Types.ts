import { NavigatorTypes } from './navigator/';
import { LoginTypes } from './login/';
import { UtilityTypes } from './utility/';
import { StorageTypes, SaveByKey, RemoveKey, GetByKey } from './storage';
import { GeocodeTypes, GeocodeAddress } from './geocode';
import { AutoLogin, Logout } from './login/actions/';

/**
 * The global application state as contained in Redux
 *
 * All state objects might not be available.
 *
 * @export
 * @interface IApplicationState
 */
export interface IApplicationState {
  route?: NavigatorTypes.INavigationState;
  login?: LoginTypes.ILoginState;
  utility?: UtilityTypes.IUtilityState;
  storage?: StorageTypes.IStorageState;
  geocoding?: GeocodeTypes.IGeocodeState;
}

/**
 * Describes possible props for a components.
 *
 * All actions and state objects might not be available.
 *
 * @export
 * @interface IProps
 */
export interface IProps extends IApplicationState {
  // AppLoadingChanged?: appActions.AppLoadingChanged;
  dispatch?: any;
  FacebookLogin?: LoginTypes.FacebookLogin;
  AppErrorChanged?: UtilityTypes.IAppErrorChangedAction;
  AppLoadingChanged?: UtilityTypes.IAppLoadingChangedAction;
  GoogleLogin?: LoginTypes.GoogleLogin;
  GeocodeAddress?: GeocodeAddress;
  BankIdLogin?: LoginTypes.BankIdLogin;
  GetByKey?: GetByKey;
  RemoveKey?: RemoveKey;
  SaveByKey?: SaveByKey;
  AutoLogin?: AutoLogin;
  Logout?: Logout;
}
