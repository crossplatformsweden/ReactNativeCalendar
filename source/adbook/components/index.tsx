import * as types from '../../Types';
import React from 'react';
// @ts-ignore - Redux base class issue
import { Agenda } from 'react-native-calendars';
import {
  View,
  Text,
  TouchableNativeFeedback,
// @ts-ignore
  TouchableHighlight,
  Platform,
} from 'react-native';
import { Dispatch, connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Book } from '../actions/index';
import Theme from '../../styles/index';
import { ComponentBase } from '../../utility';
import { Card } from 'react-native-elements';
import { Actions } from 'react-native-router-flux';
import * as bookingTypes from '../types';
import Moment from 'moment';
import { ParkingAction } from '../types/index';

interface IAdSlot {
  name: String;
  info: String;
  date: bookingTypes.ITimePeriod;
}

const RenderCard = ({ item }: { item: IAdSlot }) => (
  <Card
    // @ts-ignore
    title={item.name}
  >
    <Text>{item.date.timeStart.format('HH:mm')} - {item.date.timeEnd.format('HH:mm')}</Text>
  </Card>
);

export class AdBookBase extends React.Component<types.IProps, {}> {
  constructor(props: types.IProps) {
    super(props);
    this.onSelected = this.onSelected.bind(this);
  }

  /**
   * Update the booking with time period from selected slot
   *
   * @param {IAdSlot} item selected slot
   * @memberof AdBookBase
   */
  onSelected(item: IAdSlot) {
    // Update booking with new date
    let localBooking = this.props.adbook.booking;
    localBooking.timeperiod = item.date;
    // TODO: skriv action creator funktion som används för att initiera IBooking
    this.props.Book(localBooking, bookingTypes.ParkingConstants.BOOKING_ADDED);
  }

  render() {
    // this.state = {
    //   bookingLocal: new types.IProps.Book()
    // };
    const today = Moment('2018-03-28');
    const items = {
            '2018-03-28': new Array<IAdSlot>({
              name: 'Name here',
              info: 'Info here',
              date: {
                timeStart: today,
                timeEnd: today.add(6, 'h'),
              },
            }),
          };
    return (
      <View style={[Theme.container, Theme.paddingDefault]}>
        <Agenda
          // the list of items that have to be displayed in agenda. If you want to render item as empty date
          // the value of date key kas to be an empty array []. If there exists no value for date key it is
          // considered that the date in question is not yet loaded
          items={items}
          // specify how each item should be rendered in agenda
          renderItem={(item: IAdSlot) => {
            return (<View>
           {Platform.OS === 'ios' ? (
                <TouchableHighlight onPress={() => this.onSelected(item)}>
                  <RenderCard item={item} />
                </TouchableHighlight>
              ) : (
                <TouchableNativeFeedback onPress={() => this.onSelected(item)}>
                  <RenderCard item={item} />
                </TouchableNativeFeedback>
              ) }
            </View>);
          }}
          onDayPress={(day: Date) => {
            console.log('day pressed');
          }}
          // specify your item comparison function for increased performance
          rowHasChanged={(r1: any, r2: any) => {
            return r1.text !== r2.text;
          }}
        />
      </View>
    );
  }
}
/**
 * AdBook component. Implements ComponentBase
 */
export const AdBook = ComponentBase(AdBookBase);

const mapStateToProps = (state: types.IApplicationState) => ({
  adbook: state.adbook,
  login: state.login,
  utility: state.utility,
});
const mapDispatchToProps = (dispatch: Dispatch<types.IProps>) => ({
  dispatch,
  Book: bindActionCreators(Book, dispatch),
});

export default connect<types.IApplicationState, types.IProps>(
  mapStateToProps,
  mapDispatchToProps
  // @ts-ignore - Redux base class issue
)(AdBook);
