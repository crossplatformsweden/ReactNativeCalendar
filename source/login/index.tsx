// login/index
import LoginContainer, { LoginBase, Login } from './components';
import LoginReducer from './reducers';
import * as LoginTypes from './types';
import { FacebookLogin, GoogleLogin, AutoLogin, Logout, BankIdLogin } from './actions';

export {
  LoginContainer,
  LoginBase,
  Login,
  LoginReducer,
  LoginTypes,
  FacebookLogin,
  GoogleLogin,
  AutoLogin,
  Logout,
  BankIdLogin
};
