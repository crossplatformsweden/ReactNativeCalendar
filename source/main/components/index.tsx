import React from 'react';
import { View, Text, StyleSheet } from 'react-native';
import { connect, Dispatch } from 'react-redux';

import * as types from '../../Types';
import Theme from '../../styles';
import { ComponentBase } from '../../utility';

const styles = StyleSheet.create({
  padding: { padding: 20 },
});

export class MainBase extends React.Component<types.IProps, {}> {
  render() {
    return (
      <View style={[Theme.container, styles.padding]}>
        <Text>Home</Text>
      </View>
    );
  }
}

/**
 * Main component. Implements ComponentBase
 */
export const Main = ComponentBase(MainBase);

const mapStateToProps = (state: types.IApplicationState) => ({
  route: state.route,
  login: state.login,
  utility: state.utility,
});

const mapDispatchToProps = (dispatch: Dispatch<types.IProps>) => ({
  dispatch,
});

export default connect<types.IApplicationState, types.IProps>(
  mapStateToProps,
  mapDispatchToProps
  // @ts-ignore - Redux base class issue
)(Main);
