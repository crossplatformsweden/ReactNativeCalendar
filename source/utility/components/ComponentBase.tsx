// utility/components/ComponentBase
import * as React from 'react';
import { View } from 'react-native';
import { Actions } from 'react-native-router-flux';
import { FormValidationMessage } from 'react-native-elements';
import Sentry from 'sentry-expo';

import BusyIndicator from './BusyIndicator';
import * as types from '../../Types';
import * as NavigatorTypes from '../../navigator/types';
import Theme from '../../styles';

/**
 * Base class for componets with busy indicator and validation feedback.
 *
 * Busy indicator requires props to contain "login" from Redux store.
 *
 * Validator requires props to contain "utility" from Redux store
 *
 * @export
 * @abstract
 * @class ComponentBase
 * @extends {Component<{}, {}>}
 */
const componentBase = <P, S>(WrappedComponent: React.ComponentType<P>) =>
  class ComponentBase extends React.Component<P, S> {
    protected constructor(props: P) {
      super(props);
    }

    /**
     * Check login state and navigate if needed
     *
     * @memberof ComponentBase
     */
    protected async checkLogin() {
      const iProps = this.props as types.IProps;
      if (iProps && iProps.login && iProps.login.isLoggedIn) {
        // User is logged in - set user info for debugging
        if (iProps.login.user)
          Sentry.setUserContext({
            email: iProps.login.user.email,
            username: iProps.login.user.username,
            extra: { type: iProps.login.user.type },
          });

        if (
          Actions.currentScene === NavigatorTypes.NavigationConstants.LOGIN &&
          Actions &&
          Actions.pop
        ) {
          // Close login modal
          Actions.pop();
        }
      } else if (Actions.LoginScreen) {
        // Push login screen
        if (!iProps.login.isLoggedIn) {
          // Actions.LoginScreen();
        }
      }
    }

    componentWillMount() {
      this.checkLogin();
      if (typeof super.componentWillMount === 'function') {
        super.componentWillMount();
      }
    }

    componentWillReceiveProps(nextProps: Readonly<P>, nextContext: any) {
      this.checkLogin();
      if (typeof super.componentWillReceiveProps === 'function') {
        super.componentWillReceiveProps(nextProps, nextContext);
      }
    }

    componentDidMount() {
      if (typeof super.componentDidMount === 'function') {
        super.componentDidMount();
      }
    }

    componentDidUpdate(
      prevProps: Readonly<P>,
      prevState: Readonly<S>,
      prevContext: any
    ) {
      if (typeof super.componentDidUpdate === 'function') {
        super.componentDidUpdate(prevProps, prevState, prevContext);
      }
    }

    componentWillUnmount() {
      if (typeof super.componentWillUnmount === 'function') {
        super.componentWillUnmount();
      }
    }

    render() {
      const iProps = this.props as types.IProps;
      return (
        <View style={Theme.container}>
          <WrappedComponent {...this.props} />
          {iProps && iProps.utility && iProps.utility.hasError === true ? (
            <FormValidationMessage testID='validationMessage'>
              {iProps.utility.errorMessage}
            </FormValidationMessage>
          ) : null}
          <BusyIndicator
            testID='busyIndicator'
            isBusy={iProps.utility.isBusy}
            message={iProps.utility.busyReason}
          />
        </View>
      );
    }
  };

export default componentBase;
