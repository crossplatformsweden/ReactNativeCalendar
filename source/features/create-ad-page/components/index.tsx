// login/components/index
import React from 'react';
import { bindActionCreators } from 'redux';
import { View } from 'react-native';
import { connect, Dispatch } from 'react-redux';

import * as types from '../../../Types';
import * as Navigator from '../../../navigator/components';
import { CreateAd } from '../actions';
// @ts-ignore
import CalendarPicker from 'react-native-calendar-picker';
import moment from 'moment';

let startDate = moment();
let stopDate = moment();

export class CreateAdBooking extends React.Component<types.IProps, {}> {
  constructor(props: types.IProps) {
    super(props);
    this.onDateChange = this.onDateChange.bind(this);
  }
  pickDate = async () => {
    await this.props.createAd;
  }
  onDateChange(date: moment.Moment, type: string) {
    if (type === 'END_DATE') {
      stopDate = date;
      // console.log(this.props.createAd.createAd.fromDate);
      let localAd = this.props.createAd.createAd;
      // console.log(localAd);
      // localAd.fromDate = startDate;
      // localAd.toDate = stopDate;
      // this.props.CreateAd(localAd);
    } else {
      startDate = date;
    }
  }

  render() {
    return (
      <Navigator.ModalBase hideClose>
        <View style={{ marginTop: 100, flex: 1 }}>
          <CalendarPicker
            startFromMonday={true}
            allowRangeSelection={true}
            onDateChange={this.onDateChange}
            weekdays={['Mån', 'Tis', 'Ons', 'Tor', 'Fre', 'Lör', 'Sön']}
            selectedDayColor='#00bfff'
          />
        </View>
      </Navigator.ModalBase>
    );
  }
}

const mapStateToProps = (state: types.IApplicationState) => ({
  route: state.route,
  createAd: state.createAd,
});

const mapDispatchToProps = (dispatch: Dispatch<types.IProps>) => ({
  CreateAd: bindActionCreators(CreateAd, dispatch),
  dispatch,
});

export default connect<types.IApplicationState, types.IProps>(
  mapStateToProps,
  mapDispatchToProps
  // @ts-ignore - Redux base class issue
)(CreateAdBooking);
