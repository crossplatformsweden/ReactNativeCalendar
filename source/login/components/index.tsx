import React from 'react';
import { bindActionCreators } from 'redux';
import { connect, Dispatch } from 'react-redux';
import { View, StyleSheet, Image } from 'react-native';
import { SocialIcon } from 'react-native-elements';

import * as types from '../../Types';
import { StorageTypes, GetByKey } from '../../storage';
import { FacebookLogin, GoogleLogin } from '..';
import { ComponentBase } from '../../utility';
import { CreateAdBooking } from '../../features/create-ad-page/components/index';

interface IState {}

/**
 * Login base component. Doesn't implement ComponentBase
 * @class
 */
export class LoginBase extends React.Component<types.IProps, IState> {
  componentWillMount() {
    if (!this.props.login.isLoggedIn) {
      this.props.GetByKey(StorageTypes.StorageConstants.STORAGE_USER_KEY);
    }

    if (super.componentWillMount) {
      super.componentWillMount();
    }
  }

  /**
   * Wraps {@link FacebookActions.FacebookLogin}
   * @type {function}
   * @private
   */
  async facebookLogin() {
    await this.props.FacebookLogin();
    // @ts-ignore
    if (super.checkLogin) {
      // @ts-ignore
      super.checkLogin();
    }
  }

  render() {
    return (
      <View style={styles.container}>
        <View style={styles.logoContainer}>
          <Image
            source={require('../../../resources/icon.png')}
            style={{
              width: 120,
              height: 120,
            }}
          /></View>
        <SocialIcon
          title='Sign In With Facebook'
          button
          // @ts-ignore - id is unknown
          id='buttonFacebook'
          onPress={this.facebookLogin}
          type='facebook'
        />
        <SocialIcon
          title='Sign In With Google'
          button
          // @ts-ignore - id is unknown
          id='buttonGoogle'
          onPress={this.props.GoogleLogin}
          type='google-plus-official'
        />
        <CreateAdBooking />
      </View>
    );
  }
}

/**
 * Login component. Implements ComponentBase
 */
export const Login = ComponentBase(LoginBase);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 20,
    marginTop: 10,
  },
  logoContainer: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'flex-start',
  },
});

const mapStateToProps = (state: types.IApplicationState) => ({
  login: state.login,
  utility: state.utility,
});

const mapDispatchToProps = (dispatch: Dispatch<types.IProps>) => ({
  FacebookLogin: bindActionCreators(FacebookLogin, dispatch),
  GoogleLogin: bindActionCreators(GoogleLogin, dispatch),
  GetByKey: bindActionCreators(GetByKey, dispatch),
  dispatch,
});

export default connect<types.IApplicationState, types.IProps>(
  mapStateToProps,
  mapDispatchToProps
  // @ts-ignore - Redux base class issue
)(Login);