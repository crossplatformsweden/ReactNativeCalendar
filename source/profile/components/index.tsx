// profile/components/index
import * as React from 'react';
import { connect, Dispatch } from 'react-redux';
import { bindActionCreators } from 'redux';
import { NavigationActions } from 'react-navigation';
import { Actions } from 'react-native-router-flux';

import { View, ScrollView, StyleSheet, Switch } from 'react-native';
import { Avatar, Button, FormLabel, FormInput } from 'react-native-elements';
import _ from 'lodash';
import * as ProfileTypes from '../types';
import * as types from '../../Types';
import * as login from '../../login';
import * as storage from '../../storage';
import * as nav from '../../navigator';
import { ComponentBase } from '../../utility';
import Theme from '../../styles';

const styles = StyleSheet.create({
  logout: {
    margin: 10,
  },
  imageContainer: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'flex-start',
  },
});

interface IState {
  userLocal: login.LoginTypes.IUser;
}

/**
 * Allows the user to edit their details and settings
 * @param {module:TypesModule.DefaultReducer} storage
 */
export class ProfileBase extends React.Component<
  ProfileTypes.IProfileProps,
  IState
> {
  constructor(props: ProfileTypes.IProfileProps) {
    super(props);
    this.state = {
      userLocal: new login.LoginTypes.User(null, null, null, null),
    };

    // Ensure methods have "this" context
    this.saveUserLocalToStorage = this.saveUserLocalToStorage.bind(this);
    this.userLocalFromStorage = this.userLocalFromStorage.bind(this);
    this.userPropertyChanged = this.userPropertyChanged.bind(this);
    this.autologinChanged = this.autologinChanged.bind(this);
  }

  /**
   * Cast local storage to {@see login.LoginTypes.IUser} instance
   *
   * @returns {login.LoginTypes.IUser}
   * @memberof ProfileBase
   */
  storageValueToUser(): login.LoginTypes.IUser {
    return this.valueIsUser(this.props.storage.value)
      ? (this.props.storage.value as login.LoginTypes.IUser)
      : null;
  }

  /**
   * Determines if the value is an {@see login.LoginTypes.IUser} instance
   *
   * @param {*} value
   * @returns {value is login.LoginTypes.IUser}
   * @memberof ProfileBase
   */
  valueIsUser(value: any): value is login.LoginTypes.IUser {
    return !_.isNil(value) && !_.isNil(value.type);
  }

  /**
   * Set local state from storage
   *
   * @memberof ProfileBase
   */
  userLocalFromStorage() {
    if (
      !_.isNil(this.props.storage) &&
      this.valueIsUser(this.props.storage.value)
    ) {
      const storageUser = this.storageValueToUser();
      const newUserLocal = Object.assign({}, this.state.userLocal, storageUser);
      this.setState({ userLocal: newUserLocal });
    }
  }

  /**
   * Update user object in local state
   * @param newValue new user value
   * @memberof ProfileBase
   */
  async userPropertyChanged(newValue: any) {
    const newUserLocal = Object.assign({}, this.state.userLocal, newValue);
    // Update local user
    this.setState({
      userLocal: newUserLocal,
    });
  }

  /**
   * Save local state to isolated storage
   *
   * @memberof ProfileBase
   */
  async saveUserLocalToStorage() {
    // Update user in storage
    await this.props.SaveByKey(
      storage.StorageTypes.StorageConstants.STORAGE_USER_KEY,
      this.state.userLocal
    );

    // Update local object
    this.userLocalFromStorage();
  }

  async componentWillMount() {
    // Get user from storage
    this.userLocalFromStorage();
    if (_.isNil(this.state.userLocal) || _.isNil(this.state.userLocal.name)) {
      await this.props.GetByKey(
        storage.StorageTypes.StorageConstants.STORAGE_USER_KEY
      );
      this.userLocalFromStorage();
    }

    if (typeof super.componentWillMount === 'function') {
      super.componentWillMount();
    }
  }

  async autologinChanged() {
    let user = this.state.userLocal;
    user.autologin = !user.autologin;
    this.setState({userLocal: user});
    await this.saveUserLocalToStorage();
  }

  render() {
    return (
      <View style={[Theme.container, Theme.paddingDefault]}>
        <ScrollView testID='scrollView'>
          <View style={[Theme.container, Theme.paddingDefault]}>
            {!_.isNil(this.state.userLocal) &&
            !_.isNil(this.state.userLocal.picture) ? (
              <View>
                <View style={[styles.imageContainer, Theme.paddingDefault]}>
                  <Avatar
                    large
                    rounded
                    source={{ uri: this.state.userLocal.picture }}
                    activeOpacity={0.7}
                  />
                </View>
                <FormLabel>Namn</FormLabel>
                <FormInput
                  placeholder='Namn'
                  autoCapitalize='words'
                  value={this.state.userLocal.name}
                  onBlur={this.saveUserLocalToStorage}
                  onChangeText={val => this.userPropertyChanged({ name: val })}
                />
                <FormLabel>Användarnamn</FormLabel>
                <FormInput
                  placeholder='Användarnamn'
                  value={this.state.userLocal.username}
                  autoCapitalize='none'
                  onBlur={this.saveUserLocalToStorage}
                  onChangeText={val =>
                    this.userPropertyChanged({ username: val })
                  }
                />
                <FormLabel>Lösenord</FormLabel>
                <FormInput
                  placeholder='Lösenord'
                  secureTextEntry
                  autoCapitalize='none'
                  value={this.state.userLocal.password}
                  onBlur={this.saveUserLocalToStorage}
                  onChangeText={val =>
                    this.userPropertyChanged({ password: val })
                  }
                />
                <FormLabel>E-postadress</FormLabel>
                <FormInput
                  placeholder='E-postadress'
                  keyboardType='email-address'
                  autoCapitalize='none'
                  value={this.state.userLocal.email}
                  onBlur={this.saveUserLocalToStorage}
                  onChangeText={val => this.userPropertyChanged({ email: val })}
                />
                <FormLabel>Autologin</FormLabel>
                <Switch
                  style={{ marginLeft: 15, marginTop: 5, marginBottom: 10 }}
                  value={this.state.userLocal.autologin}
                  onValueChange={this.autologinChanged}
                />
              </View>
            ) : null}
            <Button
              // @ts-ignore
              medium
              style={styles.logout}
              icon={{ name: 'sign-out', type: 'font-awesome' }}
              title='Sign out'
              onPress={this.props.Logout}
            />
          </View>
        </ScrollView>
      </View>
    );
  }
}

export const Profile = ComponentBase(ProfileBase);

const mapStateToProps = (state: types.IApplicationState) => ({
  storage: state.storage,
  login: state.login,
  utility: state.utility,
});

const mapDispatchToProps = (
  dispatch: Dispatch<ProfileTypes.IProfileProps>
) => ({
  Logout: async () => {
    // Remove user from storage
    await dispatch(
      storage.RemoveKey(storage.StorageTypes.StorageConstants.STORAGE_USER_KEY)
    );
    // Set state to logged out
    dispatch({ type: login.LoginTypes.LoginConstants.LOGGED_OUT });
    // Display login screen
    Actions.LoginScreen();
  },
  Login: () =>
    dispatch(
      NavigationActions.navigate({
        routeName: nav.NavigatorTypes.NavigationConstants.LOGIN,
      })
    ),
  RemoveKey: bindActionCreators(storage.RemoveKey, dispatch),
  SaveByKey: bindActionCreators(storage.SaveByKey, dispatch),
  GetByKey: bindActionCreators(storage.GetByKey, dispatch),
  dispatch,
});

export default connect<ProfileTypes.IProfileProps, {}>(
  mapStateToProps,
  mapDispatchToProps
  // @ts-ignore - Redux base class issue
)(Profile);
