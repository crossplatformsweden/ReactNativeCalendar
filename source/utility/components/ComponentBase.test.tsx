// utility/components/ComponentBase.test
/// <reference types="jest"/>
import React from 'react';
import { shallow } from 'enzyme';
import { GooglePlacesAutocomplete } from 'react-native-google-places-autocomplete';
// @ts-ignore
import { MapView } from 'expo';

import * as types from '../../Types';
import { LoginTypes } from '../../login';
import { Map } from '../../map/components/MapComponent';
import BusyIndicator from './BusyIndicator';

jest.unmock('react-native');

// Navigator will call connected components
jest.mock('../../navigator', () => ({
  Routes: 'View',
  AppNavigator: 'Button',
}));

function setup() {
  const user: LoginTypes.IUser = {
    accessToken: '1',
    name: 'test',
    picture: '',
    type: 'Google',
  };
  const props: types.IProps = {
    login: { isLoggedIn: false },
    utility: { isBusy: false, busyReason: null, hasError: false },
  };

  // We'll use Map as a concrete implementation of ComponentBase
  const enzymeWrapper = shallow(<Map {...props} />);

  return {
    props,
    enzymeWrapper,
  };
}

describe('components', () => {
  describe('ComponentBase', () => {
    it('should render self and subcomponents', () => {
      const { enzymeWrapper } = setup();
      expect(enzymeWrapper).toMatchSnapshot();

      const busyProps = enzymeWrapper.find(BusyIndicator).props();
      expect(busyProps.isBusy).toBe(false);
    });

    it('should call lifecycle methods', () => {
      const componentWillMount = jest.spyOn(
        Map.prototype,
        'componentWillMount'
      );
      const componentDidMount = jest.spyOn(Map.prototype, 'componentDidMount');
      const componentWillReceiveProps = jest.spyOn(
        Map.prototype,
        'componentWillReceiveProps'
      );
      const componentDidUpdate = jest.spyOn(
        Map.prototype,
        'componentDidUpdate'
      );

      setup();

      // Test lifecycle methods
      expect(componentWillMount).toHaveBeenCalled();
      expect(componentDidMount).toHaveBeenCalled();

      // Lifecycle methods NOT called first run
      expect(componentWillReceiveProps).not.toHaveBeenCalled();
      expect(componentDidUpdate).not.toHaveBeenCalled();

      componentWillMount.mockReset();
      componentWillMount.mockRestore();
      componentDidUpdate.mockReset();
      componentDidUpdate.mockRestore();
      componentDidMount.mockReset();
      componentDidMount.mockRestore();
      componentWillReceiveProps.mockReset();
      componentWillReceiveProps.mockRestore();
    });
  });
});
