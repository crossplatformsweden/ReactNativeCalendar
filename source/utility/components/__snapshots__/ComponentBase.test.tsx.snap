// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components ComponentBase should render self and subcomponents 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ComponentBase
    login={
      Object {
        "isLoggedIn": false,
        "type": null,
      }
    }
    utility={
      Object {
        "busyReason": null,
        "hasError": false,
        "isBusy": false,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <MainBase
          login={
            Object {
              "isLoggedIn": false,
              "type": null,
            }
          }
          utility={
            Object {
              "busyReason": null,
              "hasError": false,
              "isBusy": false,
            }
          }
        />,
        null,
        <busyIndicator
          isBusy={false}
          message={null}
          testID="busyIndicator"
        />,
      ],
      "style": Object {
        "flex": 1,
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "login": Object {
            "isLoggedIn": false,
            "type": null,
          },
          "utility": Object {
            "busyReason": null,
            "hasError": false,
            "isBusy": false,
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      null,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "isBusy": false,
          "message": null,
          "testID": "busyIndicator",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <MainBase
            login={
              Object {
                "isLoggedIn": false,
                "type": null,
              }
            }
            utility={
              Object {
                "busyReason": null,
                "hasError": false,
                "isBusy": false,
              }
            }
          />,
          null,
          <busyIndicator
            isBusy={false}
            message={null}
            testID="busyIndicator"
          />,
        ],
        "style": Object {
          "flex": 1,
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "login": Object {
              "isLoggedIn": false,
              "type": null,
            },
            "utility": Object {
              "busyReason": null,
              "hasError": false,
              "isBusy": false,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "isBusy": false,
            "message": null,
            "testID": "busyIndicator",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
